### PR TITLE
Don't leak FunctionRef instances

### PR DESCRIPTION
**Pull request by drewhk**
They were not removed anywhere from the parent in streams code (they are removed in Akka typed).

### CODE REVIEW

**Comment by drewhk** [comment]
I am not sure if the fix is 100% correct. It eliminates the leak but I don't know if there are some races there (children containers and DeathWatchNotification messages), so @rkuhn you should take a look.

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 1]
```diff
@@ -178,7 +178,11 @@ object GraphStageLogic {
       behaviour = receive
     }
 
-    def stop(): Unit = functionRef.stop()
+    def stop(): Unit = {
+      functionRef.stop()
+      // Can this be here?
+      cell.removeFunctionRef(functionRef)
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: and null out the cell.. oh

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 2]
```diff
@@ -132,14 +132,14 @@ object GraphStageLogic {
                          initialReceive: StageActorRef.Receive) {
 
     private val callback = getAsyncCallback(internalReceive)
+    private val cell = materializer.supervisor match {
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: this should probably be a `def`

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 3]
```diff
@@ -178,7 +178,11 @@ object GraphStageLogic {
       behaviour = receive
     }
 
-    def stop(): Unit = functionRef.stop()
+    def stop(): Unit = {
+      functionRef.stop()
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: this now needs to go, otherwise we call it twice

**Comment by rkuhn** [comment]
LGTM apart from the comments—I don’t smell any race conditions here.

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala (hunk 3 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: ah, I see, this it calls it.

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala (hunk 2 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: right, makes sense

**Comment by drewhk** [comment]
updated

**Comment by drewhk** [comment]
I was not sure about the order between stop and remove, but since the latter calls the former, I expect the order is taken properly care of now.

**Comment by rkuhn** [comment]
Should all be good now.

**Comment by drewhk** [comment]
refs: https://github.com/akka/akka/issues/19670

**Pull request merged**